### PR TITLE
Native disk update

### PR DESCRIPTION
**Pull request by s4heid**
# Summary

This change introduces a new cloud method called `update_disk`, which can be used to update disk parameters on an existing Azure Disk. Disk size and storage account type can be changed on any data disk. For storage account types PremiumV2_LRS and Ultra Disks, iops and throughput (mbps) can also be changed.

Relates to: https://github.com/cloudfoundry/bosh/issues/2496

# Checklist:

Please check each of the boxes below for which you have completed the corresponding task:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] All unit tests pass locally (after my changes)
- [ ] Rubocop reports zero offenses (after my changes) -- there is a complexity violation in the disk manager. either we disable the check for the method or we can discuss if you prefer a different solution.

Please include below the summary portions of the output from the following 2 scripts:

  ```
  pushd src/bosh_azure_cpi
    ./bin/test-unit
    ./bin/rubocop_check
  popd
  ```

  _NOTE:_ Please see how to setup dev environment and run unit tests in docs/development.md.

### Unit Test output:

```
Finished in 6.62 seconds (files took 0.73321 seconds to load)
1044 examples, 0 failures
```

### Integration Test output:

```sh
$ bundle exec rspec ./spec/integration
...
Finished in 1 minute 17.39 seconds (files took 0.43431 seconds to load)
3 examples, 0 failures
```

### Rubocop output:

```
Inspecting 190 files
.......C........................................C.............................................................................................................................................

Offenses:

lib/cloud/azure/disk/disk_manager2.rb:48:5: C: Metrics/ParameterLists: Method has too many optional parameters. [4/3]
    def update_disk(disk_id, size = nil, storage_account_type = nil, iops = nil, mbps = nil) ...
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
lib/cloud/azure/vms/vm_manager.rb:21:5: C: Metrics/AbcSize: Assignment Branch Condition size for create is too high. [<87, 178, 53> 205.1/192]
    def create(bosh_vm_meta, location, vm_props, disk_cids, network_configurator, env, agent_settings, network_spec, config) ...
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
lib/cloud/azure/vms/vm_manager.rb:21:5: C: Metrics/CyclomaticComplexity: Cyclomatic complexity for create is too high. [45/39]
    def create(bosh_vm_meta, location, vm_props, disk_cids, network_configurator, env, agent_settings, network_spec, config) ...
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
lib/cloud/azure/vms/vm_manager.rb:21:5: C: Metrics/MethodLength: Method has too many lines. [189/184]
    def create(bosh_vm_meta, location, vm_props, disk_cids, network_configurator, env, agent_settings, network_spec, config) ...
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
lib/cloud/azure/vms/vm_manager.rb:21:5: C: Metrics/PerceivedComplexity: Perceived complexity for create is too high. [50/45]
    def create(bosh_vm_meta, location, vm_props, disk_cids, network_configurator, env, agent_settings, network_spec, config) ...
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

190 files inspected, 5 offenses detected
```

# Changelog

* The `update_disk` cloud method has been introduced, allowing updates to disk parameters for Azure Disks, including disk size and storage account type. For PremiumV2_LRS and Ultra Disks, IOPS and throughput (mbps) adjustments are now possible.

### CODE REVIEW

**Review by aramprice — APPROVED** [review]
This seems good from a structural perspective. I don't have much context on the Azure internals so it may be helpful to have input from others who have more Azure experience.

File: src/bosh_azure_cpi/lib/cloud/azure/cloud.rb [hunk 1]
```diff
@@ -479,6 +479,44 @@ def resize_disk(disk_cid, new_size)
...
+
+          caching = cloud_properties['caching']
+          raise Bosh::Clouds::NotSupported, 'Disk caching cannot be modified' if caching && (caching != disk[:caching])
+
+          new_size_in_gib = mib_to_gib(new_size)
+          old_size_in_gib = disk[:disk_size]
+          raise Bosh::Clouds::NotSupported, 'Disk size cannot be decreased' if old_size_in_gib > new_size_in_gib
+
+          new_size_in_gib = nil if new_size_in_gib == old_size_in_gib
+          account_type = cloud_properties['storage_account_type']
```
Comment (MSSedusch) [inline src/bosh_azure_cpi/lib/cloud/azure/cloud.rb:505]: @s4heid do you want to implement the snapshot based migration if the storage type is chaged from Premium V2/UltraSSD -> Premium v1/Standard SSD/Standard HDD?

Comment (s4heid) [reply src/bosh_azure_cpi/lib/cloud/azure/cloud.rb:505]: Yes, I think the snapshot-based storage type change is a useful feature. I have planned to do some testing next week and eventually provide a pull request if everything goes well.

Comment (s4heid) [reply src/bosh_azure_cpi/lib/cloud/azure/cloud.rb:505]: I don't see it as a requirement for this change, though. The Azure API returns a meaningful error when trying to switch from PremiumV2 to a lower tier storage account.

**Review by MSSedusch — APPROVED** [review]
lgtm

**Comment by rkoster** [comment]
Thanks! @s4heid

**Pull request merged**